File: packages/spotify_sdk_ios/ios/spotify_sdk_ios/Classes/RemoteManager.swift
import Flutter
import SpotifyiOS

class RemoteManager: NSObject {
    static let shared = RemoteManager()

    var appRemote: SPTAppRemote?
    var connectionStatusHandler: ConnectionStatusHandler?
    var playerStateHandler: PlayerStateHandler?
    var playerContextHandler: PlayerContextHandler?
    var capabilitiesHandler: CapabilitiesHandler?
    var userStatusHandler: UserStatusHandler?

    static var playerStateChannel: FlutterEventChannel?
    static var playerContextChannel: FlutterEventChannel?
    static var capabilitiesChannel: FlutterEventChannel?
    static var userStatusChannel: FlutterEventChannel?
    static var connectionStatusChannel: FlutterEventChannel?

    override init() {
        super.init()
    }
}

File: packages/spotify_sdk_android/android/src/main/kotlin/de/minimalme/spotify_sdk/RemoteManager.kt
package de.minimalme.spotify_sdk

import android.app.Activity
import android.content.Context
import com.spotify.android.appremote.api.SpotifyAppRemote
import de.minimalme.spotify_sdk.subscriptions.ConnectionStatusChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.event.SetEvent
import kotlinx.event.event

class RemoteManager {

    var applicationContext: Context? = null
    var applicationActivity: Activity? = null
    var spotifyAppRemote: SpotifyAppRemote? = null

    var playerContextChannel: EventChannel? = null
    var playerStateChannel: EventChannel? = null
    var capabilitiesChannel: EventChannel? = null
    var userStatusChannel: EventChannel? = null
    var connectionStatusChannel: EventChannel? = null

    var connStatusEventChannel: SetEvent<ConnectionStatusChannel.ConnectionEvent> = event()
    var pendingOperation: PendingOperation? = null

    inline fun withAppRemote(result: Result, block: (SpotifyAppRemote) -> Unit) {
        val remote = spotifyAppRemote
        if (remote != null && remote.isConnected) {
            block(remote)
        } else {
            result.error("spotifyAppRemoteNull", "spotifyAppRemote is null or disconnected", "")
        }
    }

    fun String.checkAndSetPendingOperation(result: Result) {
        if (pendingOperation == null) {
            pendingOperation = PendingOperation(this, result)
        } else {
            result.error("errorPendingOperation", "Operation $this cannot be started because another operation is pending: ${pendingOperation?.methodName}", "")
        }
    }
}
